dnf-behave-tests/features/steps/common/sqlite_repodata.py

```diff
@@ -0,0 +1,134 @@
+                "checksum_type"]
+
+def load_sqlite(sqlite_path, repodata_type):
+    metadata_obj = Metadata()
+    con = sqlite3.Connection(sqlite_path)
+    metadata_cursor = con.cursor()
+
+    if repodata_type == "primary":
+        parse_pkg = parse_primary_pkg_sqlite
+    elif repodata_type == "filelists":
+        parse_pkg = parse_filelists_pkg_sqlite
+    elif repodata_type == "other":
+        parse_pkg = parse_other_pkg_sqlite
+
+    metadata_cursor.execute("SELECT * FROM packages")
+    for row in metadata_cursor:
+        p = parse_pkg(row, con)
+        metadata_obj.append(p.checksum, p)
+    return metadata_obj
+
+def parse_primary_pkg_sqlite(row, con):
+    pp = PrimaryPackage()
+    uid = row[0]
+    for key, val in zip(COL_PACKAGES, row[1:]):
+        setattr(pp, key, val)
+    if pp.location_base is None:
+        pp.location_base = ''
+
+    cur = con.cursor()
+    # provides, conflicts, obsoletes
+    for pco in ("provides", "conflicts", "obsoletes"):
+        pco_set = set([])
+        cur.execute("SELECT * FROM %s WHERE pkgKey=?" % pco, (uid,))
+        for name, flag, epoch, ver, rel, _ in cur:
+            pco_set.add((name, flag, epoch, ver, rel))
+        setattr(pp, pco, pco_set)
+
+    # requires
+    req_set = set([])
+    cur.execute('SELECT * FROM requires WHERE pkgKey=?', (uid,))
+    for name, flag, epoch, ver, rel, _, pre in cur:
+        if pre == 'TRUE':
+            pre = True
+        else:
+            pre = False
+        req_set.add((name, flag, epoch, ver, rel, pre))
+    setattr(pp, 'requires', req_set)
+
+    # files
+    files = set([])
+    dirs = set([])
+    ghosts = set([])
+    cur.execute('SELECT name, type FROM files WHERE pkgKey=?', (uid,))
+    for filename, ftype in cur:
+        if ftype == 'file':
+            files.add(filename)
+        elif ftype == 'dir':
+            dirs.add(filename)
+        elif ftype == 'ghost':
+            ghosts.add(filename)
+    pp.files = files
+    pp.dirs = dirs
+    pp.ghosts = ghosts
+    return pp
+
+def parse_filelists_pkg_sqlite(row, con):
+    fp = FilelistsDbPackage()
+    pkgkey = row[0]
+    fp.checksum = row[1]
+    cur = con.cursor()
+    cur.execute("SELECT dirname, filenames, filetypes FROM filelist WHERE pkgKey=?", (pkgkey,))
+
+    for dirname, filenames, filetypes in cur:
+        # XXX: If there is dir without name '/' it's splited into 2 empty string
+        #      so, if there are two empty string next to each other, replace
+        #      it with only one empty string
+        splited_filenames = filenames.split('/')
+        filtered_splited_filenames = []
+        if len(filenames):
+            filtered_splited_filenames.append(splited_filenames[0])
+        for i in range(1, len(filenames.split('/'))):
+            if splited_filenames[i] == '' and splited_filenames[i - 1] == '':
+                continue
+            filtered_splited_filenames.append(splited_filenames[i])
+        # XXX: End
```

Comment:
This part looks strange. I do not understand what `If there is dir without name '/' it's splited into 2 empty string` means. At first sight stripping trailing '/' before filenames.split('/') should be enough?

Reply:
I think that would not work, sqlite can have a row with filenames column like this: `//testa/testb` with dirname just `/` this is caused by rpm which has just "/" in its %files section. 
We want the leading single "/" to result in one empty string that is than used with the dirname `/`. To get the xml equivalent `<file type="dir">/</file>`

I think this is a special case for the '/' because there is no better way how to express it with the dirname, filenames schema.

Reply:
I think I managed to simplify the code in the newest version.